### PR TITLE
Adds runtime warning on divide-by-zero, fixes use of "out" and "where" in arctan2

### LINKED ISSUE #5132: floor division by zero doesn't match numpy
I noticed the following discrepancy:
```python
In [4]: x = ak.array([0,1])
In [5]: x//0
Out[5]: array([0 0])

In [6]: x.to_ndarray() //0
<ipython-input-6-a666e56723de>:1: RuntimeWarning: divide by zero encountered in floor_divide
  x.to_ndarray() //0
```

### LINKED ISSUE #5149: ak.arctan2 uses the "where" parameter incorrectly.
Update:

This is true of every function that uses "where" in arkouda/numpy/numeric.py.

The numpy counterparts of all of these functions only use "where" if "out" is also used.  We don't have "out" in any of our functions.

--------

ak.arctan2 uses "where" incorrectly.

np.arctan2 only uses "where" if the "out" parameter is also specified.  If "where" is True, it puts arctan2(a,b) in "out."  If "where" is False, it leaves the "out" parameter unchanged.  So, for example:

```
a = np.arange(17)
b = np.arange(17) - 8
c = np.ones(17)
np.arctan2(b,a,where=False)
array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])

```

Instead, ak.arctan2 returns arctan2(a,b) when "where" is True, and a/b when "where" is False.

```
za = ak.array(a)
zb = ak.array(b)
ak.arctan2(zb,za,where=False).to_ndarray()
array([       -inf, -7.        , -3.        , -1.66666667, -1.        ,
       -0.6       , -0.33333333, -0.14285714,  0.        ,  0.11111111,
        0.2       ,  0.27272727,  0.33333333,  0.38461538,  0.42857143,
        0.46666667,  0.5       ])

```

This can cause divide-by-zeros.  In an upcoming PR where we generate RuntimeWarnings on divide-by-zero, this causes test_arctan2 to fail, because both arctan2 and test_arctan2 divide by denom (which can include zeros) when where is False.

I suggest eliminating the "where" parameter from ak.arctan2, since it depends on "out," and we don't incorporate the "out" parameter anywhere in arkouda that I know of.

### PR DESCRIPTION
**Pull request by drculhane**
Closes #5132.  Closes #5149.

- The binops in pdarrayclass.py now give a RuntimeWarning or a Floating Point Error (or nothing) if / or // cause a divide-by-zero, _depending on the "divide" setting of ak.errstate_.

- Because the RuntimeWarning surfaced an issue in `arctan2`, I've rewritten that function somewhat.  It now has an "out" parameter matching numpy, and uses the "where" parameter correctly.

- Most of the original arctan2 function remains as `_arctan2_`, because (_subjective opinion here_) I think that calling it as a separate function improves the readability of the code.

- In doing this rewriting, I removed some mypy ignores that are no longer needed.

- I rewrote the unit tests for `arctan2` to use our newer assert functions.  I think this cleans up that code significantly.  In the cases that use "where," I check both the returned result, and the returned "out" pdarray, since they should match.

Update:

Since other tests were also causing divide-by-zero errors, I extended the divide-by-zero check to use `ak.errstate`.  I also bypassed additional tests in datetime once a divide-by-zero occurred, because they would also cause errors.

### CODE REVIEW

File: arkouda/numpy/numeric.py [hunk 1]
```diff
@@ -1409,11 +1413,13 @@ def arctan2(
...
-    num : pdarray or numeric_scalars
+    x1 : pdarray or numeric_scalars
         Numerator of the arctan2 argument.
-    denom : pdarray or numeric_scalars
+    x2 : pdarray or numeric_scalars
         Denominator of the arctan2 argument.
-    where : bool or pdarray, default=True
+    out : Optional, pdarray
+        A pdarray in which to store the result, or to use as a source when where is False.
+    where : Optional, bool or pdarray, default=True
```
[the remark this answers is not in the document]
Comment (ajpotts) [reply arkouda/numpy/numeric.py]: See also "Default set to True."  later in the docstring.

[the remark this answers is not in the document]
Comment (ajpotts) [reply arkouda/numpy/numeric.py]: Also, where should accept `bool_scalars`, not just `bool`.  That would match what NumPy allows.  Also, this will change your error statements below.

File: arkouda/numpy/numeric.py [hunk 2]
```diff
@@ -1441,72 +1447,99 @@ def arctan2(
...
-        if where is True:
-            return ret
-        else:
-            new_pda = num / denom  # type : ignore
-            return _merge_where(new_pda, where, ret)
+        return create_pdarray(repMsg)
 
-    else:
-        return scalar_array(arctan2(num, denom) if where else num / denom)
+    else:  # should be impossible to reach here.
```
Comment (ajpotts) [inline arkouda/numpy/numeric.py]: I don't like the comment "should be impossible to reach here.".  If it's truly impossible it should throw an error.  If it's not impossible, the comment should be removed.

Comment (ajpotts) [reply arkouda/numpy/numeric.py]: Actually, what I like better is allow scalar/scalar to the inputs, as mentioned in my previous comment.

**Review by ajpotts — APPROVED** [review]
Great work!

File: arkouda/numpy/numeric.py (hunk 2 above)
Comment (drculhane) [reply arkouda/numpy/numeric.py]: This is a helper function (underscore arctan2 underscore), which will never be called if x1 and x2 are both scalars.  Given that you want the scalar case handled, I'll put that in the arctan2 function itself.

File: arkouda/numpy/numeric.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (drculhane) [reply arkouda/numpy/numeric.py]: afaik, numpy arctan2 doesn't allow bool_scalars in the where parameter.  See below.
```

>>> x1
array([0., 1., 2., 3., 4.])
>>> x2
array([-3., -2., -1.,  0.,  1.])
>>> out
array([1., 1., 1., 1., 1.])
>>> wint
array([-4, -3, -2, -1,  0])
>>> wflt
array([-4., -3., -2., -1.,  0.])
>>> wbin
array([ True,  True,  True,  True, False])
>>> np.arctan2(x1,x2,out,where=wflt)
Traceback (most recent call last):
  File "<python-input-26>", line 1, in <module>
    np.arctan2(x1,x2,out,where=wflt)
    ~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^
TypeError: Cannot cast array data from dtype('float64') to dtype('bool') according to the rule 'safe'
>>> np.arctan2(x1,x2,out,where=wint)
Traceback (most recent call last):
  File "<python-input-27>", line 1, in <module>
    np.arctan2(x1,x2,out,where=wint)
    ~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^
TypeError: Cannot cast array data from dtype('int64') to dtype('bool') according to the rule 'safe'
>>> np.arctan2(x1,x2,out,where=wbin)
array([3.14159265, 2.67794504, 2.03444394, 1.57079633, 1.        ])
```

File: arkouda/numpy/numeric.py [hunk 3]
```diff
@@ -1409,11 +1413,13 @@ def arctan2(
 
     Parameters
     ----------
-    num : pdarray or numeric_scalars
+    x1 : pdarray or numeric_scalars
         Numerator of the arctan2 argument.
-    denom : pdarray or numeric_scalars
+    x2 : pdarray or numeric_scalars
```
Comment (ajpotts) [inline arkouda/numpy/numeric.py]: Also, if it's a big change, may not belong in this PR, but NumPy allows bools for `x1` and `x2`: 

```python
In [5]: np.arctan2([True,False],False,where=True)
Out[5]: array([1.57, 0.  ], dtype=float16)
```

File: tests/numpy/datetime_test.py [hunk 4]
```diff
@@ -148,10 +149,20 @@ def test_op_types(self, verbose=pytest.verbose):
...
                 assert isinstance(ret, return_type)
                 metrics["ak_supported"] += 1
+                # try:
+                #     compare_flag = True
+                #     ret = eval(f"fcvec {op} scvec")
+                #     assert isinstance(ret, return_type)
+                #     metrics["ak_supported"] += 1
+                # except RuntimeWarning:
+                #     continue  # this test can cause divide-by-zero, which would
+                #     # also cause a following test to fail, so skip it
```
[the remark this answers is not in the document]
Comment (drculhane) [reply tests/numpy/datetime_test.py]: Ooops.  Done.

File: arkouda/numpy/numeric.py [hunk 5]
```diff
@@ -1441,72 +1460,210 @@ def arctan2(
...
+
+    #  Begin with various checks.
 
-    if not all(isSupportedNumber(arg) or isinstance(arg, pdarray) for arg in [num, denom]):
+    #  First the scalar case.
+
+    if np.isscalar(x1) and np.isscalar(x2):
+        if out is None:
+            return nparctan2(x1, x2) if where is None or where is True else np.float64(1.0)
+
```
[the remark this answers is not in the document]
Comment (ajpotts) [reply arkouda/numpy/numeric.py]: In NumPy, the `where=` argument on ufuncs like `arctan2` is a masked write, not a conditional computation. When `where=False` and no `out=` array is provided, NumPy skips writing the result but still returns an output object whose memory is left uninitialized. For scalar inputs, this means the returned value is undefined and may appear to change across calls due to memory reuse. Because this behavior is subtle, unsafe, and easy to misuse, Arkouda should reject this pattern explicitly: if where contains any False values and no out parameter is supplied, Arkouda should raise an error rather than returning nondeterministic results. This enforces deterministic semantics and prevents silent data corruption in distributed and pandas-compatible workflows.
